lib/usagereporter/teleport/aggregating/service.go

```diff
@@ -70,6 +72,93 @@ func prepareUserActivityReport(
 	return report, nil
 }
 
+// resourcePresenceReportKey returns the backend key for a resource presence report with
+// a given UUID and start time, such that reports with an earlier start time
+// will appear earlier in lexicographic ordering.
+func ResourcePresenceReportKey(reportUUID uuid.UUID, startTime time.Time) []byte {
+	return backend.Key(ResourcePresenceReportsPrefix, startTime.Format(time.RFC3339), reportUUID.String())
+}
+
+// prepareResourcePresenceReport prepares a resource presence report for storage.
+// It returns a slice of reports for case when single report is too large to fit in a single item.
+// As even a single resource kind report can be too large to fit in a single item, it may return
+// multiple reports with single resource kind report split into multiple fragments.
+func prepareResourcePresenceReports(
+	clusterName, reporterHostID []byte,
+	startTime time.Time, records []*prehogv1.ResourceKindPresenceReport,
+) ([]*prehogv1.ResourcePresenceReport, error) {
+	reports := make([]*prehogv1.ResourcePresenceReport, 0, 1) // at least one report
+
+	for len(records) > 0 {
+		reportUUID := uuid.New()
+		report := &prehogv1.ResourcePresenceReport{
+			ReportUuid:          reportUUID[:],
+			ClusterName:         clusterName,
+			ReporterHostid:      reporterHostID,
+			StartTime:           timestamppb.New(startTime),
+			ResourceKindReports: records,
+		}
+
+		if proto.Size(report) <= maxItemSize {
+			// The last report fits, so we're done.
+			reports = append(reports, report)
+			return reports, nil
+		}
+
+		// We're over the size limit, so we need to split the report and try again.
+		report.ResourceKindReports = make([]*prehogv1.ResourceKindPresenceReport, 0, len(records)-1)
+
+		emptyKindReportSize := proto.Size(&prehogv1.ResourceKindPresenceReport{
+			ResourceKind: records[0].GetResourceKind(),
+			ResourceIds:  make([]uint64, 1), // need to have non-empty slice to calculate size
+		})
+		singleItemSize := int(unsafe.Sizeof(uint64(0)))
```

Comment:
Don't bring unsafe in just to calculate a constant value.
```suggestion
		const singleItemSize = 8
```